portfolio/src/main/java/com/google/sps/data/BlogHashMap.java

```diff
@@ -0,0 +1,199 @@
+  // Returned values contain the most recent messages.
+  public LinkedList<BlogMessage> getMessages(int limit) {
+    if (limit < 0) {
+      return new LinkedList<>();
+    }
+
+    Set<String> keys = map.keySet();
+    LinkedList values = new LinkedList();
+
+    for (String key : keys) {
+      if (limit <= 0) {break;}
+      Iterator iterator;
+      int tagMessagesSize = (map.get(key)).size(); 
+
+      // startPoint is from (n0.of.messages in each tag - limit) if the list contains more messages
+      // than the limit,
+      // That way, returned values contain the most recent messages.
+      if (tagMessagesSize > limit) {
+        int startPoint = tagMessagesSize - limit;
+        iterator = (map.get(key)).listIterator(startPoint);
+      } 
+      // Otherwise the n0.of messages in the tag is less than or equal to the limit, hence
+      // we return all of them.
+      else {
+        iterator = (map.get(key)).iterator();
+      }
+      while(iterator.hasNext() && limit >= 0){
+        values.addLast(iterator.next());
+        limit--;
+      }
+    }
+    return values;
+  }
+
+  // Takes a tag, and
+  // Returns a linkedList of all values of BlogMessage type for the specified tag.
+  public LinkedList<BlogMessage> getMessages(String tag) {
+    LinkedList<BlogMessage> values = new LinkedList<BlogMessage>();
+    if (map.containsKey(tag)) {
+      values.addAll(map.get(tag));
+    } else {
+      System.err.println("Tag not found. Tag : " + tag);
+    }    
+    return values;
+  }
+
+  // Returns a requested amount of all BlogMessages for a specified tag.
+  public LinkedList<BlogMessage>getMessages(String tag, int limit) {
+    if (limit < 0) {
+      return new LinkedList<>();
+    }
+
+    LinkedList values = new LinkedList();
+
+    if (map.containsKey(tag)){
+      Iterator iterator;
+      int tagMessagesSize = (map.get(tag)).size();
+      if (tagMessagesSize > limit) {
+        int startPoint = tagMessagesSize - limit;
+        iterator = (map.get(tag)).listIterator(startPoint);
+      } else {
+        iterator = (map.get(tag)).iterator();
+      }
+      while (iterator.hasNext() && limit >= 0) {
+        values.addLast(iterator.next());
+        limit--;
+      }
+    } else {
+      System.err.println("Tag not found. Tag: " + tag);
+    }
+    return values;
+  } 
+
+  // Takes a list of tags, and
+  // Returns a linkedList of all values of BlogMessage type for the specified tags.
+  public LinkedList<BlogMessage> getMessages(List<String> tags) {
+    LinkedList<BlogMessage> values = new LinkedList<BlogMessage>();
+    for (String tag : tags){
+      if (map.containsKey(tag)) {
+        values.addAll(map.get(tag));
+      } else {
+        System.err.println("Tag not found. Tag : " + tag);
+      }    
+    }
+    return values;
+  }
+
+  // Returns a requested amount of all BlogMessages for specified tags.
+  public LinkedList<BlogMessage> getMessages(List<String> tags, int limit) {
```

Comment:
You only really need this function. If tags is empty or null, get all tags. If limit <= 0 don't use a limit. Just document that in the comments of the function. This way you reuse a lot of code and you limit the number of functions that need to be tested.

This also reduces the PR size to a more reasonable level.

Reply:
Sounds good!

Reply:
I like how you organized this with the private getMessages functions serving internal purposes!